PhysicsTools/PatAlgos/python/jetCollectionTools.py

```diff
@@ -0,0 +1,621 @@
+
+    if inputCollection and inputCollection not in [
+        "slimmedJets", "slimmedJetsAK8", "slimmedJetsPuppi", "slimmedCaloJets",
+      ]:
+      raise RuntimeError("Invalid input collection: %s" % inputCollection)
+
+    if bTagDiscriminators is None:
+      bTagDiscriminators = self.bTagDiscriminators
+
+    if JETCorrLevels is None:
+      JETCorrLevels = self.JETCorrLevels
+    
+    #
+    # Decide which jet collection we're dealing with
+    #
+    jetLower = jet.lower()
+    jetUpper = jet.upper()
+    tagName = jetUpper
+    recoJetInfo = RecoJetInfo(jet, inputCollection)
+
+    if inputCollection == "slimmedJets":
+      assert(jetLower == "ak4pfchs")
+    elif inputCollection == "slimmedJetsAK8":
+      assert(jetLower == "ak8pfpuppi")
+    elif inputCollection == "slimmedJetsPuppi":
+      assert(jetLower == "ak4pfpuppi")
+    elif inputCollection == "slimmedCaloJets":
+      assert(jetLower == "ak4calo")
+    
+    #=======================================================
+    #
+    # If jet collection in MiniAOD is  not 
+    # specified, build the jet collection.
+    #
+    #========================================================
+    if not inputCollection or recoJetInfo.doCalo:
+      print("custom_jme_cff::RecoJetAdder::addRecoJetCollection: inputCollection in NanoAOD not specified. Building recojet collection now")
+
+      #=======================================================
+      #
+      # Prepare the inputs to jet clustering
+      #
+      #========================================================
+      #
+      # Set up PF candidates
+      #
+      pfCand = self.pfLabel
+      #
+      # Setup PU method for PF candidates
+      # 
+      if recoJetInfo.jetPUMethod not in [ "", "cs" ]:
+        pfCand += recoJetInfo.jetPUMethod
+      #
+      #
+      #
+      if pfCand not in self.prerequisites:
+        #
+        # Skip if no PU Method or CS specified
+        #
+        if recoJetInfo.jetPUMethod in [ "", "cs" ]:
+          pass
+        #
+        # CHS
+        #
+        elif recoJetInfo.jetPUMethod == "chs":
+          setattr(proc, pfCand,
+            cms.EDFilter("CandPtrSelector",
+              src = cms.InputTag(self.pfLabel),
+              cut = cms.string("fromPV"),
+            )
+          )
+          self.prerequisites.append(pfCand)
+        #
+        # PUPPI
+        #
+        elif recoJetInfo.jetPUMethod == "puppi":
+          setattr(proc, pfCand,
+            puppi.clone(
+              candName   = cms.InputTag(self.pfLabel),
+              vertexName = cms.InputTag(self.pvLabel),
+            )
+          )
+          self.prerequisites.append(pfCand)
+        #
+        # Softkiller
+        #
+        elif recoJetInfo.jetPUMethod == "sk":
+          setattr(proc, pfCand,
+            softKiller.clone(
+              PFCandidates = cms.InputTag(self.pfLabel),
+              rParam       = cms.double(recoJetInfo.jetSizeNr),
```

Comment:
please drop type specifications for all parameters which already exist.
This is a safer way to protect from parameter name mistakes and will also help in possible parameter migrations.